### DISCUSSION TITLE
Cannot not receive realtime updates when running Pocketbase behind nginx with an SSL let's encrypt certificate.

### DISCUSSION THREAD

**Discussion by paulsutherland** [pocketbase/pocketbase#480] [Q&A] [answered]
Hello,

I am having an issue with receiving realtime subscription updates when running Pocketbase behind nginx with an SSL let's encrypt certificate.

I have an Ubuntu 22.04 server on digital ocean. I have everything setup correctly - smtp, digital ocean spaces for files etc and it works without any issues. (Great job - this is a fantastic high quality piece of work btw)

If I run Pocketbase directly with:

`./Pocketbase serve --http="example.com:80" --https="example.com:443"`

All is well.  The client app can CRUD and subscribe to live changes.

If I put it in front of nginx and simply use this block:

```
server {
    listen 80;
    server_name example.com;

    location / {
        # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 180s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8090;
    }
}
```

again it works without any issues.

If I run it with this block:

```
server {

  server_name example.com;

  location  / {
    # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
    proxy_set_header Connection '';
    proxy_http_version 1.1;
    proxy_read_timeout 180s;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;

    proxy_pass http://127.0.0.1:8090;
  }

  listen [::]:443 ssl ipv6only=on; # managed by Certbot
  listen 443 ssl; # managed by Certbot
  ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem; # managed by Certbot
  ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem; # managed by Certbot
  include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
  ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}
server {
    if ($host = example.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

  listen 80;
  listen [::]:80;

  server_name example.com;
    return 404; # managed by Certbot
}
```

Then the front end can CRUD records, but it won't receive any realtime updates.  I can't figure out why and I have tried many different configurations.

Is there anything obvious that I am missing by setting up nginx and ssl in front of Pocketbase?

FYI, the front end code is just out of the API docs:

```
const client = new PocketBase('https://example.com')

client.realtime.subscribe('chats', function (e) {
    console.log(e.record)
})
```

I do see these timeout errors in the nginx error logs:

```
2022/09/15 21:04:10 [error] 2401#2401: *140 upstream timed out (110: Unknown error) while reading upstream, client: MY_IP_ADDRESS, server: example.com, request: "GET /api/realtime HTTP/1.1", upstream: "http://127.0.0.1:8090/api/realtime", host: "example.com", referrer: "http://localhost:8888/"

2022/09/15 21:09:41 [error] 2401#2401: *164 upstream timed out (110: Unknown error) while reading upstream, client: MY_IP_ADDRESS, server: example.com, request: "GET /api/realtime HTTP/1.1", upstream: "http://127.0.0.1:8090/api/realtime", host: "example.com", referrer: "http://localhost:8888/"
```

Thanks in advance,
Paul

**Comment by ganigeorgiev** [comment, marked as answer]
I haven't tested extensively the SSE integration behind a reverse proxy, but the above maybe is caused by the default buffering that nginx applies (but if that is really the case, then I'm not sure why it works with the non-ssl version).

Could you try to disable the buffering by setting `proxy_buffering off`?

_If that works, maybe it will be a good idea to add a `X-Accel-Buffering: no` as a response header in the code so that the buffering can be managed individually per endpoint as explained in https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering._

**Reply by paulsutherland** [reply]
@ganigeorgiev 

Adding `proxy_buffering off;` done the trick.  It works as expected now.

The final block that works for serving Pocketbase behind nginx with an SSL let's encrypt certificate is:

```
server {

  server_name example.com;

  location  / {
    # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
    proxy_set_header Connection '';
    proxy_http_version 1.1;
    proxy_read_timeout 180s;

    proxy_buffering off;   # <<< Solved the problem
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;

    proxy_pass http://127.0.0.1:8090;
  }

  listen [::]:443 ssl ipv6only=on; # managed by Certbot
  listen 443 ssl; # managed by Certbot
  ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem; # managed by Certbot
  ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem; # managed by Certbot
  include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
  ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}
server {
    if ($host = example.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

  listen 80;
  listen [::]:80;

  server_name example.com;
    return 404; # managed by Certbot
    
}
```
Thank you for the quick response and best of luck with the project. 

Cheers...

Paul

**Comment by paulsutherland** [comment]
@ganigeorgiev 

It is worth noting that turning off the proxy buffering generates an error message every 3-4 min in the browser console.  Realtime still works but overtime it floods the console.

<img width="924" alt="Screenshot 2022-09-16 at 09 34 10" src="https://user-images.githubusercontent.com/2683905/190596855-5fa21754-5ddc-40b9-b51a-787df5323ea3.png">

Reverting to `./Pocketbase serve --http="example.com:80" --https="example.com:443"` the console error disappears.

Thanks,

Paul

**Reply by ganigeorgiev** [reply]
Hm, that's most likely because of the `proxy_read_timeout` is less than what the server-side is using (currently it is 5min - https://github.com/pocketbase/pocketbase/blob/master/apis/realtime.go#L74). I'll update later the docs to match.

_btw. I've just pushed a [v0.7.3 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.7.3) that adds the `X-Accel-Buffering: no` so disabling globally the buffering should no longer be necessarily._

**Reply by paulsutherland** [reply]
@ganigeorgiev 

Yes thanks, our message and your release crossed paths.  I tested 0.7.3 and removed proxy_buffering off; from nginx.  It works fine, but the same error appears in the browser console with that release.

Cheers...

Paul

**Reply by ganigeorgiev** [reply]
@paulsutherland If the error is returned periodically after some inactivity, then probably the culprit is the `proxy_read_timeout`. I can't find in the nginx docs how to disable it (https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout), but you can try setting it to something large (eg. [`1M` for one month](https://nginx.org/en/docs/syntax.html)) and leave the pending requests timeout to the PocketBase server to handle.

But even if you keep it as it is, it is ok because the SSE client has autoreconnect support and the error shouldn't matter too much for the end-user experience.

_I'll try to squeeze some time after the refactoring in #376 to test the reverse proxy setup more thoroughly._

**Reply by paulsutherland** [reply]
@ganigeorgiev 
Yes of course, I realise from your posts your are head down busy on #376.

I set the proxy_read_timeout to 99999 as Nginx fell over with 1M 😄 It works for now.  I'll look into the Nginx docs to see if I can find out the limits of the configuration.

**Reply by osseonews** [reply]
Can you please share how you got this running on Digital Ocean? I don't see any docker image here, so I'm not even sure how to get started. The project seems fantastic.

**Reply by ganigeorgiev** [reply]
@osseonews There is nothing specific to DigitalOcean when deploying your application. In fact, PocketBase is completely portable and you don't need Docker or even nginx/apache. Just spin a new droplet, upload the [latest PocketBase linux amd64 binary](https://github.com/pocketbase/pocketbase/releases) and create a systemd to keep the process up as shown in the production guide or in https://github.com/pocketbase/pocketbase/discussions/498#discussioncomment-3662208.

If you need further help with the deployment, feel free to start a new discussion.

**Comment by paulsutherland** [comment]
@osseonews 

There is probably a better place to put this, but for completeness in this discussion I'll start with here 😉 

[The documentation](https://pocketbase.io/docs/going-to-production/) was more than good enough for me to get up and running with self hosting Pocketbase.

Allow me to expand on the previous answer and share some resources I find helpful, and which I return to when creating a droplet on Digital Ocean. (Other hosting providers are available, but DO is quite quick to get up and running and the droplets are good value for money. I use them when evaluating and testing out new software).

**Create a droplet**
1. Create an account at [digital ocean](https://www.digitalocean.com/).
2. Create an Ubuntu 22.04 droplet on a basic plan and the CPU option of regular with SSD to get started.
3. Setup your droplet: [Server Setup with Ubuntu 22.04](https://www.digitalocean.com/community/tutorials/initial-server-setup-with-ubuntu-22-04).

**Download Pocketbase binary**
Once you have the droplet configured as per the DO docs, login to the droplet and then download and install Pocketbase.
1. Login and navigate to your home directory
` cd ~/`
2. Install wget so you can download the Pocketbase base binary.
`sudo apt-get install wget` 
3. [Download the latest Pocketbase release](https://github.com/pocketbase/pocketbase/releases) `wget https://github.com/pocketbase/pocketbase/releases/download/v0.7.3/pocketbase_0.7.3_linux_amd64.zip` 
4. Create a directory in your home directory for Pocketbase.
`mkdir database`
5. Install unzip to allow you to unzip the archive
`sudo apt-get install unzip`
6. Extract the binary into the database directory 
`unzip pocketbase_0.7.3_linux_amd64.zip -d database`
6. Run Pocketbase from the command line 
`./database/pocketbase serve --http="Your Droplet IP Address:80"`

You now have Pocketbase running and you can visit the admin page in your browser to begin the setup: `http://Your Droplet IP Address`

**Adding a custom domain**
If you want to serve Pocketbase over https, you need a custom domain name. It won't work on a bare ip address.

If you want to use a custom domain, then you should register a domain and configure your DNS to point to your digital ocean droplet IP address.  Here is an example using Gandi.net with an A Record.

<img width="496" alt="Screenshot 2022-09-17 at 12 22 29" src="https://user-images.githubusercontent.com/2683905/190855061-418fe1bd-c2b8-4155-bbbb-b29c6183edeb.png">

If you have a custom domain setup, you can let Pocketbase handle the setup of https and serving your instance over your custom domain.

`./database/serve --http="Your Droplet IP Address:80" --https="Your Droplet IP Address:443"`

You should complete this by creating a systemd service, as [explained in the docs](https://pocketbase.io/docs/going-to-production/#:~:text=Create%20Systemd%20service) so that Pocketbase can run as a service and autostart when the server restarts.  

**Optional** Nginx reverse proxy.

It is a matter of preference to use a reverse proxy, but it is something I have done for years.  I prefer to run Nginx in front of services as it is a battle hardened web server and I don't like exposing smaller custom apps directly on the web. 

As previously answered you don't need to do this, and giving the rapid pace of development and the fact that the project has not yet reached a stable v1.0 release, I would recommend sticking with Pocketbase serving on its own for now. There is still some tuning required to get Nginx and Pocketbase working perfectly as highlighted in this discussion.  I just wanted to complete the circle for myself.

For completeness, the following are helpful resources:

1. [Setup Nginx on a Ubuntu 22.04 droplet.](https://www.digitalocean.com/community/tutorials/how-to-install-nginx-on-ubuntu-22-04)
2. [Secure Nginx with Let's Encrypt on Ubuntu 22.04](https://www.digitalocean.com/community/tutorials/how-to-secure-nginx-with-let-s-encrypt-on-ubuntu-22-04)

The Nginx config is explained in the docs, but for reference here is what works for me:

```
server {

  server_name <your-subdomain.your-customdomain.com>;

  location / {
    # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
    proxy_set_header Connection '';
    proxy_http_version 1.1;
    proxy_read_timeout 999999;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-NginX-Proxy true;

    proxy_pass http://127.0.0.1:8090;
  }

  listen [::]:443 ssl ipv6only=on; # managed by Certbot
  listen 443 ssl; # managed by Certbot
  ssl_certificate /etc/letsencrypt/live/<your-subdomain.your-customdomain.com>/fullchain.pem; # managed by Certbot
  ssl_certificate_key /etc/letsencrypt/live/<your-subdomain.your-customdomain.com>/privkey.pem; # managed by Certbot
  include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
  ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}
server {
    if ($host = <your-subdomain.your-customdomain.com>) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

  listen 80;
  listen [::]:80;

  server_name <your-subdomain.your-customdomain.com>;
    return 404; # managed by Certbot

}
```

With the above setup, your systemd config will differ from the example in the docs:

```
[Unit]
Description = pocketbase

[Service]
Type           = simple
User           = YOUR_USER     << the user you setup ubuntu with is fine as you are not binding to port 80 or 443, otherwise it is root
Group          = YOUR_GROUP    << the user you setup ubuntu with is fine as you are not binding to port 80 or 443, otherwise it is root
LimitNOFILE    = 4096
Restart        = always
RestartSec     = 5s
StandardOutput = append:/your/path/to/logs/errors.log
StandardError  = append:/your/path/to/logs/errors.log
ExecStart      = /home/<Username>/database/pocketbase serve

[Install]
WantedBy = multi-user.target
```

This overview maybe a bit rough, but I hope that is helpful. Good luck.

Cheers...
Paul

**Comment by osseonews** [comment]
Thanks! Extremely helpful.

**Comment by osseonews** [comment]
Follow up: I was able to get Pocketbase running on DO with the instructions above in about 5 minutes. So thanks very much! I didn't do the custom domain yet, but this is enough to start testing using pocketbase, and I'll try custom domain later. 

Just a few notes to help others. On DO, unzip is not installed so, you need to run `apt install unzip` before step 5. Also, step 6 didn't work for me as written above for some reason. What did work is after unzipping going into the database directory (cd database) and running this: `./pocketbase serve --http="YOURDROPLETIP:80" ` - I think step 6 is just missing pocketbase, so maybe it should be: ./database/pocketbase serve --http="Your Droplet IP Address:80"

**Reply by paulsutherland** [reply]
@osseonews 

It has made the effort to document my installation worth it if you managed to get it up and running in such a short space of time, as it can be quite a journey to pull it all together. To be fair - I find the DO docs to be first class and the Pocketbase docs are equally good to follow.

Thanks for the feedback, I pulled it together from my shell history, so I am not surprised a few steps were not quite right.

I have updated the instructions regarding this two issues you highlighted.

Cheers....
Paul
